numpy/core/numeric.py

```diff
@@ -458,10 +465,31 @@ class ndarray is returned.
     >>> np.asanyarray(a) is a
     True
 
+    Make sure the result is compatible with a specified dtype
+
+    >>> a = np.ones(2, dtype=np.int16)
+    >>> np.asarray(a, precision=np.float32)
+    array([ 1.,  1.], dtype=float32)
+    >>> a = np.ones(2, dtype=np.complex128)
+    >>> np.asarray(a, precision=np.float32)
+    array([ 1.+0.j,  1.+0.j])
+    >>> a = np.ones(2, dtype=np.complex64)
+    >>> np.asarray(a, precision=0.0)
+    array([ 1.+0.j,  1.+0.j], dtype=complex64)
+
     """
-    return array(a, dtype, copy=False, order=order)
+    if precision is not None and dtype is not None:
+        raise ValueError('cannot specify both dtype and precision')
+
+    ret = array(a, dtype, copy=False, order=order)
```

Comment:
Agreed. Somewhat similarly, rather than testing only for `precision is None`, one could avoid a needless `astype` by:
```
if precision is None or result_type(ret.dtype, precision) == ret.dtype:
    return ret
else:
    return ret.astype(precision, order=order, copy=False)
```
(it seems `np.can_cast` might be slightly faster than the comparison using `result_type` above).